### PR TITLE
Add support for opening entity selection dialogs

### PR DESCRIPTION
**Pull request by jelz**
Open questions:

- do we want to call this namespace "tools"?
- do we want to resolve differently depending on the `multiple` option? When selecting a single entity we resolve with an entity object and when selecting multiple entities we resolve with an array of entity objects
- do we want to reject when a user dismisses the modal window? We could also resolve with `null` or an empty array

Current implementation resolves differently and rejects when the window is closed without selecting something, but it felt bad when I worked with it.

Now I think I'd prefer to always resolve with an array:
- `[{single: true}]`
- `[{multiple: 1}, {multiple: 2}]`
- and `[]` when dismissed

Input highly appreciated.

### CODE REVIEW

**Comment by geigerzaehler** [comment]
> do we want to call this namespace "tools"?

“tools” seems to generic. Maybe “dialog”? @hlabas?

> do we want to resolve differently depending on the multiple option? When selecting a single entity we resolve with an entity object and when selecting multiple entities we resolve with an array of entity objects

I think we should always resolve to an array.

> do we want to reject when a user dismisses the modal window? We could also resolve with null or an empty array

I’m for returning an empty array.

**Comment by jelz** [comment]
As discussed today, the final API looks like this:

```
extension.dialogs = {
  // returned promise resolves with an object (or null if skipped)
  selectSingleEntry,
  // returned promise resolves with an object (or null if skipped)
  selectSingleAsset,
  // returned promise resolves with an array of objects (or null if skipped)
  selectMultipleEntries,
  // returned promise resolves with an array of objects (or null if skipped)
  selectMultipleAssets
}
```

File: docs/ui-extensions-sdk-frontend.md [hunk 1]
```diff
@@ -270,3 +271,58 @@ Listens for DOM changes and calls `updateHeight()` when the size changes.
...
+// display a dialog for selecting a single entry
+dialogs.selectSingleEntry().then((selectedEntry) => {})
+
+// select a single entry of "blogpost" content type
+// and display result in "de-DE" locale
+dialogs.selectSingleEntry({
+  locale: 'de-DE',
+  contentTypes: ['blogpost']
+}).then((selectedEntry) => {})
+~~~
```
Comment (geigerzaehler) [inline docs/ui-extensions-sdk-frontend.md]: Please add _In upcoming release_. See the [developer docs](https://github.com/contentful/ui-extensions-sdk/blob/master/DEVELOPING.md) for more info

File: CHANGELOG.md [hunk 2]
```diff
@@ -3,6 +3,17 @@ All notable changes to this project will be documented in this file.
 This project adheres to [Semantic Versioning](http://semver.org/) and
 the [“Keep a Changelog” format](http://keepachangelog.com/).
 
+
+## 3.1.0 - 2016-11-15
```
Comment (geigerzaehler) [inline CHANGELOG.md]: Use “Upcoming” as the title. See the [DEVELOPING.md](https://github.com/contentful/ui-extensions-sdk/blob/master/DEVELOPING.md).

File: docs/ui-extensions-sdk-frontend.md [hunk 3]
```diff
@@ -270,3 +271,58 @@ Listens for DOM changes and calls `updateHeight()` when the size changes.
...
+
+##### `dialogs.selectSingleEntry(options)`
+Opens a dialog for selecting a single entry. It returns a promise that is
+resolved with the selected entity or `null` if a user closes the dialog without
+selecting anything.
+
+`options` is an optional object configuring the dialog. Available `options`:
+
+* `locale` - code of a locale that you want to use. It'll be used to display
+  proper titles and descriptions. Defaults to the default space's locale
```
Comment (geigerzaehler) [inline docs/ui-extensions-sdk-frontend.md]: “Defaults to the space's default locale”

File: lib/api/dialogs.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
+export default function createDialogs (channel) {
+  return {
+    selectSingleEntry: call.bind(null, 'Entry', false),
+    selectSingleAsset: call.bind(null, 'Asset', false),
+    selectMultipleEntries: call.bind(null, 'Entry', true),
+    selectMultipleAssets: call.bind(null, 'Asset', true)
+  }
+
+  function call (entityType, multiple, options) {
+    options = Object.assign({}, options, {entityType, multiple})
```
Comment (geigerzaehler) [inline lib/api/dialogs.js]: `Object.assign()` is [not supported by IE](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign#Browser_compatibility). SAD. I personally use the [`object-assign`](https://www.npmjs.com/package/object-assign) package.

File: CHANGELOG.md [hunk 5]
```diff
@@ -3,6 +3,18 @@ All notable changes to this project will be documented in this file.
...
+## Upcoming
+### Added
+This version introduces a new functionality:
+
+- `extension.dialogs.select...(options)` opens a dialog for selecting entries
+  and assets
+
+You can now open the native entity selection dialog window from your extension.
+`options` object passed to the call allows to configure behavior of search and
+insertion. Please see API reference for more details.
```
Comment (ChrisChinchilla) [inline CHANGELOG.md]: @jelz Which API? This one?

https://www.contentful.com/developers/docs/references/content-management-api/#/reference/editor-interface

Comment (jelz) [inline CHANGELOG.md]: @ChrisChinchilla API reference in this repo: https://github.com/contentful/ui-extensions-sdk/blob/master/docs/ui-extensions-sdk-frontend.md

**Pull request merged**